# Preproceccing

Removing n.a.

In [43]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
realty_data = pd.read_csv('data.csv',index_col=0)

In [44]:
realty_df = realty_data.copy()
#realty_df.columns = ['id','number_of_rooms','type','metro','adress','size','buinding','parking','price','telephone','description',]

In [45]:
realty_df['Город'] = realty_df['Адрес'].str.split(',').str[0]
realty_df = realty_df[realty_df['Город']=='Москва']
realty_df.drop('Город',axis=1, inplace=True)

In [46]:
realty_df.head()

ID  объявления Количество комнат       Тип                         Метро  \
0       271271157                 4  Квартира  м. Смоленская (9 мин пешком)   
1       271634126                 4  Квартира  м. Смоленская (8 мин пешком)   
2       271173086   4, Оба варианта  Квартира  м. Смоленская (7 мин пешком)   
3       272197456   4, Оба варианта  Квартира  м. Смоленская (3 мин пешком)   
4       273614615                 2  Квартира   м. Арбатская (7 мин пешком)   

                              Адрес      Площадь, м2  \
0     Москва, улица Новый Арбат, 27       200.0/20.0   
1     Москва, улица Новый Арбат, 27  198.0/95.0/18.0   
2     Москва, улица Новый Арбат, 27  200.0/116.0/4.0   
3  Москва, переулок Плотников, 21С1  170.0/95.0/17.0   
4     Москва, улица Новый Арбат, 15    58.0/38.0/5.0   

                         Дом   Парковка  \
0           5/16, Монолитный  подземная   
1  5/16, Монолитно-кирпичный  подземная   
2                       5/16  подземная   
3                        5/6  подземная   
4           12/26, Панельный        NaN   

                                                Цена  \
0  500000.0 руб./ За месяц, Залог - 500000 руб., ...   
1  500000.0 руб./ За месяц, Залог - 500000 руб., ...   
2  500000.0 руб./ За месяц, Залог - 500000 руб., ...   
3  400000.0 руб./ За месяц, Залог - 400000 руб., ...   
4  225000.0 руб./ За месяц, Залог - 225000 руб., ...   

                     Телефоны  ...             Окна  \
0                +79166369231  ...              NaN   
1                +79850608590  ...  На улицу и двор   
2  +79672086536, +79099269384  ...  На улицу и двор   
3  +79660342340, +79099269384  ...  На улицу и двор   
4                +79852432860  ...  На улицу и двор   

                           Санузел           Можно с детьми/животными  \
0                              NaN  Можно с детьми, Можно с животными   
1  Совмещенный (2), Раздельный (1)                     Можно с детьми   
2                  Совмещенный (3)                     Можно с детьми   
3                  Совмещенный (3)                  Можно с животными   
4                  Совмещенный (2)                                NaN   

                                       Дополнительно        Название ЖК  \
0  Мебель в комнатах, Мебель на кухне, Ванна, Душ...  Новый Арбат, 2010   
1  Мебель в комнатах, Мебель на кухне, Ванна, Душ...        Новый Арбат   
2  Мебель в комнатах, Мебель на кухне, Ванна, Душ...        Новый Арбат   
3  Мебель в комнатах, Мебель на кухне, Ванна, Душ...                NaN   
4  Мебель в комнатах, Мебель на кухне, Ванна, Душ...           The Book   

  Серия дома Высота потолков, м                Лифт Мусоропровод  \
0        NaN                3.0  Пасс (4), Груз (1)           Да   
1        NaN                3.5  Пасс (1), Груз (1)          Нет   
2        NaN                3.2            Пасс (1)          NaN   
3        NaN                3.2            Пасс (1)          NaN   
4        NaN                3.9  Пасс (1), Груз (1)           Да   

                      Ссылка на объявление  
0  https://www.cian.ru/rent/flat/271271157  
1  https://www.cian.ru/rent/flat/271634126  
2  https://www.cian.ru/rent/flat/271173086  
3  https://www.cian.ru/rent/flat/272197456  
4  https://www.cian.ru/rent/flat/273614615  

[5 rows x 24 columns]

In [47]:
realty_df.isna().sum()

ID  объявления                  0
Количество комнат             535
Тип                             0
Метро                         346
Адрес                           0
Площадь, м2                     0
Дом                             0
Парковка                    11174
Цена                            0
Телефоны                        0
Описание                        0
Ремонт                       2463
Площадь комнат, м2           7228
Балкон                       6630
Окна                         5150
Санузел                      2041
Можно с детьми/животными     4915
Дополнительно                 272
Название ЖК                 15281
Серия дома                  17646
Высота потолков, м           9202
Лифт                         4192
Мусоропровод                 8007
Ссылка на объявление            0
dtype: int64

**1. Количество комнат**

1041 пропущенное значение. 

Запись состоит из числа комнат и типа - смежный/изолированный. Тип отсутсвует в более чем половине записей. Восстановить его по другим данным невожможно, предлагаем опустить, оставив только число комнат.

Для заполнения пропущенных значений, изучим зависимость площади квартиры и количества комнат:

In [48]:
realty_df['Количество комнат'] = realty_df['Количество комнат'].apply(lambda x: int(x[0]) if type(x) is str else x)
realty_df['Площадь, м2'] = realty_df['Площадь, м2'].apply(lambda x: float(x[:x.find('/')]) if type(x) is str else x)
fig = px.scatter(x=realty_df['Площадь, м2'], y=realty_df['Количество комнат'])
fig.show()


На графике заметно, что площадь квартиры и количество комнат взяимосвязвно

In [49]:
realty_df[realty_df['Количество комнат'].isna()]['Площадь, м2'].sort_values()

1556      10.0
11245     11.0
6313      11.0
6282      12.0
9559      12.0
         ...  
14382     70.0
14259     75.0
10622     78.0
6751      86.3
757      120.0
Name: Площадь, м2, Length: 535, dtype: float64

In [50]:
rooms = pd.DataFrame()
rooms['mean'] = realty_df.groupby('Количество комнат')['Площадь, м2'].mean()
rooms['q1'] = realty_df.groupby('Количество комнат')['Площадь, м2'].quantile(q=0.25)
rooms['median'] = realty_df.groupby('Количество комнат')['Площадь, м2'].median()
rooms['q3'] = realty_df.groupby('Количество комнат')['Площадь, м2'].quantile(q=0.75)
rooms

mean      q1  median      q3
Количество комнат                                    
1.0                 36.989809   34.00    38.0   40.00
2.0                 54.325638   45.00    52.0   60.00
3.0                 88.677437   68.00    80.0  102.00
4.0                149.257765  115.00   145.0  180.00
5.0                215.217417  175.00   206.0  240.00
6.0                343.264844  257.25   330.0  379.25

Посчитав средние, медианы и межквартильные интервалы площади для квартир с различным количеством комнат, мы можем предположить, что если площадь квартиры попадает в интервал, то скорее всего, в ней соотвествующее количетсво комнат. 

Квартили при этом не пересекаются. В таких случаях в качестве границы будем брать среднее между верхним и нижним. Самая большая квартира без указания числа комнат попадает в категорию 4-ех комнатных. 

In [51]:
def get_room_num(size): 
    if size < 42.5: return 1
    elif size < 64: return 2
    elif size < 108.5: return 3
    else: return 4
guess_sizes = realty_df[realty_df['Количество комнат'].isna()]['Площадь, м2'].apply(get_room_num)
realty_df['Количество комнат'].fillna(guess_sizes, inplace=True)
realty_df['Количество комнат'].value_counts()

Количество комнат
2.0    7431
1.0    7157
3.0    3676
4.0    1012
5.0     333
6.0     128
Name: count, dtype: int64

**2. Метро**

346 пропущенных значений. 

Запись состоит из станциии и оценки расстояния. Необходимо разделить на два разных признака. 

Восстановление станции метро возможно по ардесу. 

In [41]:
realty_df[realty_df['Метро'].isna()]['Адрес']

189                          Москва, улица Новый Арбат, 29
233      Москва, 9-я Северная линия, 1К1, ш. Дмитровско...
244      Москва, бульвар Академика Ландау, 3, ш. Дмитро...
267      Москва, Дмитровское шоссе, 165Дк6, ш. Дмитровс...
745                      Москва, улица Академика Янгеля, 2
                               ...                        
21970                         Москва, Медынская улица, 2К1
21971                    Москва, Булатниковская улица, 5К5
22160                          Москва, Стрелецкая улица, 7
23273    Москва, Боровский проезд, 24, ш. Боровское (4 ...
23317    Москва, улица Богданова, 6к1, ш. Боровское (3 ...
Name: Адрес, Length: 346, dtype: object

**3. Ремонт**

2463 пропущенных значений. 

В датасете 4 категории для данного признака. По признаку "без ремонта" циан выдает квартиры, где он вполне себе есть. То есть под данной категорией не подразумевается отсутсвие обоев и т.п. Исходя это этого пропущенные записи мы будем относить к категории "без ремонта", так как эта кагерория соответсвует тому, что публикующий объявление решил его не указывать.  

In [52]:
realty_df['Ремонт'].value_counts()

Ремонт
Косметический    7361
Евроремонт       7004
Дизайнерский     2773
Без ремонта       136
Name: count, dtype: int64

In [54]:
realty_df['Ремонт'].fillna('не указано', inplace=True)

In [57]:
def get_monthly_payment(price):
    payment = float(price[:price.find(' ')])
    currency = price[price.find(' ')+1:price.find('/')]
    return payment, currency
realty_df['monthly_payment'], realty_df['currency'] = zip(*realty_df['Цена'].apply(get_monthly_payment))
fig = px.box(x=realty_df['monthly_payment'], y=realty_df['Ремонт'])
fig.show()

**4. Площадь комнат**

7228 пропущенных значений. 

Данные грязные, есть записи, где количество чисел площади комнат не соотвествует количеству комнат, и записи, в которых формат записи не соотвествует числовому формату. 
Поскольку данные о площади содержатся в столбце "Площадь", предлагаем удалить "Площадь комнат" из датасета 

In [36]:
realty_df[['ID  объявления','Площадь, м2','Количество комнат','Площадь комнат, м2']].head(15)

ID  объявления  Площадь, м2  Количество комнат Площадь комнат, м2
0        271271157        200.0                4.0                NaN
1        271634126        198.0                4.0        25 25 20 25
2        271173086        200.0                4.0                NaN
3        272197456        170.0                4.0        14-42-20-19
4        273614615         58.0                2.0              20 18
5        274837728         92.0                3.0                NaN
6        273643908        213.0                5.0     43-25-25-25-22
7        274475342         98.0                3.0           26 22 15
8        273973191        120.0                3.0           45 25 25
9        272900409         90.0                4.0        16 14 16 20
10       271036964        170.0                4.0        16-40-20-18
11       273999490         80.0                3.0                NaN
12       274017037        100.0                3.0                 80
13       273231024        180.0                5.0                180
14       274839211        222.0                5.0                NaN